### PR TITLE
Feature/getx favorites advanced list

### PR DESCRIPTION
**Pull request by RomanovaPrime**

### CODE REVIEW

File: examples/favorites_advanced/getx_favorites_advanced/lib/base/common_bindings/home_binding.dart [hunk 1]
```diff
@@ -8,5 +13,11 @@ class HomeBinding implements Bindings {
     Get.put(NavbarController());
     // ignore: cascade_invocations
     Get.put(FavoritePuppiesController());
+    // ignore: cascade_invocations
+    Get.put(PuppiesRepository(ImagePicker()));
```
Comment (StanevPrime) [inline examples/favorites_advanced/getx_favorites_advanced/lib/base/common_bindings/home_binding.dart:17]: @RomanovaPrime is there are reason to not use cascade invocation

File: examples/favorites_advanced/getx_favorites_advanced/lib/feature_puppy/controllers/puppy_extra_details_controller.dart [hunk 2]
```diff
@@ -0,0 +1,49 @@
+import 'dart:async';
+// import 'package:flutter/cupertino.dart';
+import 'package:getx_favorites_advanced/feature_puppy/controllers/puppy_list_controller.dart';
+import 'package:get/get.dart';
+
+import 'package:favorites_advanced_base/models.dart';
+import 'package:favorites_advanced_base/repositories.dart';
+
+class PuppyExtraDetailsController extends GetxController {
+  final _repository = Get.find<PuppiesRepository>();
```
Comment (StanevPrime) [inline examples/favorites_advanced/getx_favorites_advanced/lib/feature_puppy/controllers/puppy_extra_details_controller.dart:10]: @RomanovaPrime what about introducing DI here? Thus, writing test will be easier as you will be able to inject a mock repository ;)

File: examples/favorites_advanced/getx_favorites_advanced/lib/feature_puppy/controllers/puppy_extra_details_controller.dart [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+      return;
+    }
+    lastFetchedPuppies.assignAll(await _repository.fetchFullEntities(
+        filterPuppies.map((element) => element.id).toList()));
+    // lastFetchedPuppies.forEach((element) {
+    //   debugPrint(element.name);});
+    // debugPrint(
+    //     'second print *** lastFetchedPuppies
+    //     count = ${lastFetchedPuppies.length}');
+    Get.find<PuppyListController>()
```
Comment (StanevPrime) [inline examples/favorites_advanced/getx_favorites_advanced/lib/feature_puppy/controllers/puppy_extra_details_controller.dart:46]: @RomanovaPrime, hmmm I'm not sure what are the best practices in GetX, but in general, the implicit dependencies are something we should avoid as much as possible. Here I would have `PuppyListController` as an explicit dependency injected via the constructor of the PuppyExtraDetailsController.

File: examples/favorites_advanced/getx_favorites_advanced/lib/feature_puppy/controllers/puppy_list_controller.dart [hunk 4]
```diff
@@ -0,0 +1,66 @@
+import 'package:flutter/material.dart';
+import 'package:get/get.dart';
+
+import 'package:favorites_advanced_base/repositories.dart';
+import 'package:favorites_advanced_base/models.dart';
+
+class PuppyListController extends GetxController {
+  PuppyListController(){
+    _repository = Get.find<PuppiesRepository>();
```
Comment (StanevPrime) [inline examples/favorites_advanced/getx_favorites_advanced/lib/feature_puppy/controllers/puppy_list_controller.dart:9]: The same as above. It's recommended converting this simplicity dependency to an implicit one

File: examples/favorites_advanced/getx_favorites_advanced/lib/feature_puppy/controllers/puppy_list_controller.dart [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    _puppies.assignAll(puppies);
+    // throw Exception('Error');
+    }catch(e){
+      _setError();
+      print(e.toString());
+    }
+    _hideLoading();
+  }
+
+  void _showLoading(){
```
Comment (StanevPrime) [inline examples/favorites_advanced/getx_favorites_advanced/lib/feature_puppy/controllers/puppy_list_controller.dart:39]: @RomanovaPrime don't you have any warning on your IDE to convert this to expression body?

File: examples/favorites_advanced/getx_favorites_advanced/lib/feature_puppy/search/views/search_page.dart [hunk 6]
```diff
@@ -1,30 +1,45 @@
...
+    final controller = Get.find<PuppyListController>();
+    return Obx(() => controller.isLoading.value
+        ? LoadingWidget()
+        : controller.hasError.value
+            ? const ErrorRetryWidget(
+                onReloadTap: _onReload,
+              )
+            : Obx(
+                () => ListView.builder(
+                  key: const Key('LoadingWidget'),
```
Comment (StanevPrime) [inline examples/favorites_advanced/getx_favorites_advanced/lib/feature_puppy/search/views/search_page.dart:20]: @RomanovaPrime I believe we have Keys as a class. So please use it here.

**Review by StanevPrime — CHANGES_REQUESTED** [review]

File: examples/favorites_advanced/getx_favorites_advanced/lib/base/common_bindings/home_binding.dart (hunk 1 above)
Comment (RomanovaPrime) [reply examples/favorites_advanced/getx_favorites_advanced/lib/base/common_bindings/home_binding.dart]: No, there isn't. I use it now.

File: examples/favorites_advanced/getx_favorites_advanced/lib/feature_puppy/controllers/puppy_extra_details_controller.dart (hunk 2 above)
Comment (RomanovaPrime) [reply examples/favorites_advanced/getx_favorites_advanced/lib/feature_puppy/controllers/puppy_extra_details_controller.dart:10]: Corrected

File: examples/favorites_advanced/getx_favorites_advanced/lib/feature_puppy/controllers/puppy_extra_details_controller.dart (hunk 3 above)
Comment (RomanovaPrime) [reply examples/favorites_advanced/getx_favorites_advanced/lib/feature_puppy/controllers/puppy_extra_details_controller.dart:46]: Corrected

File: examples/favorites_advanced/getx_favorites_advanced/lib/feature_puppy/controllers/puppy_list_controller.dart (hunk 4 above)
Comment (RomanovaPrime) [reply examples/favorites_advanced/getx_favorites_advanced/lib/feature_puppy/controllers/puppy_list_controller.dart:9]: Corrected

File: examples/favorites_advanced/getx_favorites_advanced/lib/feature_puppy/controllers/puppy_list_controller.dart (hunk 5 above)
Comment (RomanovaPrime) [reply examples/favorites_advanced/getx_favorites_advanced/lib/feature_puppy/controllers/puppy_list_controller.dart]: Actually not, but I refactored the code

File: examples/favorites_advanced/getx_favorites_advanced/lib/feature_puppy/search/views/search_page.dart (hunk 6 above)
Comment (RomanovaPrime) [reply examples/favorites_advanced/getx_favorites_advanced/lib/feature_puppy/search/views/search_page.dart:20]: I've created a new Keys file in my base folder, where to put keys that are not included in base project. When we merge my branch to develop, I can move them in base project.

File: examples/favorites_advanced/getx_favorites_advanced/lib/feature_puppy/controllers/puppy_list_controller.dart (hunk 4 above)
Comment (StanevPrime) [reply examples/favorites_advanced/getx_favorites_advanced/lib/feature_puppy/controllers/puppy_list_controller.dart:9]: @RomanovaPrime, my expectation is that PuppyListController will have PuppiesRepository as an argument

File: examples/favorites_advanced/getx_favorites_advanced/lib/feature_puppy/search/views/search_page.dart (hunk 6 above)
Comment (StanevPrime) [reply examples/favorites_advanced/getx_favorites_advanced/lib/feature_puppy/search/views/search_page.dart:20]: Didn't get that, can you elaborate?

**Review by StanevPrime — CHANGES_REQUESTED** [review]

File: examples/favorites_advanced/getx_favorites_advanced/lib/feature_puppy/search/views/search_page.dart [hunk 7]
```diff
@@ -1,30 +1,47 @@
...
+                  );
+                },
+              ),
+            ),
           ),
-          ElevatedButton(
-            onPressed: Get.find<FavoritePuppiesController>().incrementCount,
-            child: const Text('Increment Favorites'),
+          onError: (error) => ErrorRetryWidget(
+            onReloadTap: () {
```
Comment (StanevPrime) [inline examples/favorites_advanced/getx_favorites_advanced/lib/feature_puppy/search/views/search_page.dart:41]: @RomanovaPrime here you can use expression body. Also please check the static analysis is configured properly. This should be caught by it

File: examples/favorites_advanced/getx_favorites_advanced/pubspec.yaml [hunk 8]
```diff
@@ -18,18 +18,20 @@ publish_to: 'none' # Remove this line if you wish to publish to pub.dev
...
-  get: ^3.26.0
-  get_instance: ^3.13.0
-
+  appbar_textfield: ^2.0.0
+  badges: ^2.0.0-nullsafety.1
+  curved_navigation_bar: ^0.4.0-nullsafety.0
+  get: ^4.1.1
+  rxdart: ^0.26.0
+  favorites_advanced_base:
+  mockito: ^5.0.3
```
Comment (StanevPrime) [inline examples/favorites_advanced/getx_favorites_advanced/pubspec.yaml:33]: @RomanovaPrime, are you sure you need mockito as a dependency? To mee looks like we need it as `dev_dependencies` .

**Review by StanevPrime — CHANGES_REQUESTED** [review]

File: examples/favorites_advanced/getx_favorites_advanced/lib/feature_puppy/controllers/puppy_list_controller.dart [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+  List<Puppy> searchedPuppies() => [..._puppies];
+
+  Future<void> onReload() async {
+    await _initPuppies();
+    _mediatorController.clearFetchedExtraDetails();
+  }
+
+  void updatePuppiesWithExtraDetails(List<Puppy> puppiesToUpdate) {
+    puppiesToUpdate.toList().forEach((newPuppyData) {
```
[the remark this answers is not in the document]
Comment (RomanovaPrime) [reply examples/favorites_advanced/getx_favorites_advanced/lib/feature_puppy/controllers/puppy_list_controller.dart]: No reason. Changed.

File: examples/favorites_advanced/getx_favorites_advanced/lib/feature_puppy/search/views/search_page.dart (hunk 7 above)
Comment (RomanovaPrime) [reply examples/favorites_advanced/getx_favorites_advanced/lib/feature_puppy/search/views/search_page.dart:41]: - prefer_expression_function_bodies  is on, but this is not caught

File: examples/favorites_advanced/getx_favorites_advanced/pubspec.yaml (hunk 8 above)
Comment (RomanovaPrime) [reply examples/favorites_advanced/getx_favorites_advanced/pubspec.yaml:33]: I found it here https://pub.dev/packages/mockito/install

**Review by StanevPrime — APPROVED** [review]

**Pull request merged**